pkg/distributors/contracts/MultiRewardsAuthorization.sol

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20Permit.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/Authentication.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IAuthorizer.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IBasePool.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IVault.sol";
+
+/**
+ * @dev Base authorization layer implementation for MultiRewards
+ */
+abstract contract MultiRewardsAuthorization is Authentication {
+    IVault private immutable _vault;
+
+    mapping(IERC20 => mapping(IERC20 => mapping(address => bool))) private _allowlist;
+
+    constructor(IVault vault) {
+        _vault = vault;
+    }
+
+    modifier onlyAllowlistedRewarder(IERC20 pool, IERC20 rewardsToken) {
+        require(isAllowlistedRewarder(pool, rewardsToken, msg.sender), "only accessible by allowlisted rewarders");
+        _;
+    }
+
+    modifier onlyAllowlisters(IERC20 pool) {
+        require(
+            _canPerform(getActionId(msg.sig), msg.sender) ||
+                msg.sender == address(pool) ||
```

Comment:
```suggestion
            _canPerform(getActionId(msg.sig), msg.sender) ||
                msg.sender == address(pool) ||
```

Depending on how often we expect the pool to be the one allowlisting rewarders we could swap these two lines for some gas savings. Reasonably minor however.

Reply:
I suspect that the most frequent case will be balancer multisig allowlisting itself, asset managers allowlisting themselves, balancer governance allowlisting airdrops, then pool-governance allowlisting (which I don't think would be for a while since there is currently no mechanism to do so in the deployed pools.

Since this is a one-off, I'm not sure I see much savings in letting the pool contract skip the line to save gas :thinking:

Reply:
Agreed that this is very small, let's leave it as is.